### ISSUE TITLE
Rust bindings for MapLibre native

### ISSUE THREAD

**Issue by nyurik** [maplibre/maplibre-gl-native#1057]
There is currently no way to use MapLibre native from a Rust program. I think we should build Rust bindings, similar to node-js and others. 

Eventually, this may be replaced by the MapLibre-rs once it is feature complete, but this maybe a good interim solution.

**Comment by ntadej** [comment]
Yes, but I suspect we then also need proper platform support for all OSes.

**Comment by nyurik** [comment]
Lets separate concerns. For now, all platform bindings live here, so the additional platforms should be added here as well. If we ever split them up, then sure, it will be moved at the same time. Let's not do "partial fragmentation" of some projects being separate, while others are together.

If I was maintaining native, I would prefer the monorepo for these types of work, but I do not, so not my call.

**Comment by ntadej** [comment]
We put new platforms in separate repositories now, especially if they interface C++ directly and do not need anything else.

**Comment by nyurik** [comment]
@ntadej if some of the platforms are already in new repos (kinda weird for consistency reasons IMO), than sure.

**Comment by ntadej** [comment]
It depends, we now have a proposal for Xamarin where they do not need to touch the core to provide bindings. It depends on  how much platform-specific stuff you want to replace with rust implementation in rust platform.

Also there was more or less a conclusion to separate bindings out of the core repository (motivated by Metal developments as it's now hard to develop on macOS for example). I'm separating out the Qt bindings as a test/example.

In my opinion it may be useful to wait a few weeks before starting a big endeavour to see what we learn from the exercise. Your opinion is also welcome here: https://github.com/maplibre/maplibre-gl-native/pull/789

**Comment by nyurik** [comment]
@ntadej sure, but again - all this is orthogonal to the ticket -- it doesn't matter where we put the rust bindings - what's important is that we have a proper place to develop it, test it, release it, etc.

**Comment by ntadej** [comment]
I'm starting with Qt. I have an initial prototype done, but I want to make it build reliably before I open a PR.

The rest will follow when relevant people want them split and have time to do it.

**Comment by acalcutt** [comment]
I don't know about others, but one thing I like about the mono repo is it keeps knowledgeable users all in one place. As someone with little c++ skill I have always needed the opinions from contributors of other platform....which I feel would be diminished if they were just looking at the platform they care about.

**Comment by ntadej** [comment]
We should move the platforms organisation to https://github.com/maplibre/maplibre-native/pull/789

It is important to stress platform vs SDK/bindings difference (some people do not like SDK in this context, but maybe it helps here): Individual platform support will still stay in the main repository and keep the feature-parity. What is splitting is all wrappers, additional classes etc. that are not directly related to the core rendering.

**Comment by CommanderStorm** [comment]
https://github.com/maplibre/maplibre-native-rs works well enough for tile servers for now.

TODOS remaining:
- For interactive usecases we need to likely redesign or add a new API. I don't know what would be needed, issues on maplibre-native-rs more than welcome
- for more performance we will need a bit of design work
- I am not quite sure that the API we provide is very polished, so there in the short term will be a bit of breaking changes to get to a nicer place
